### ISSUE TITLE
PHP Memory overflow

### ISSUE THREAD

**Issue by presence35** [ccxt/ccxt#3226]
Hi. I'm sure I'm doing something wrong but I'm unable to figure out why I keep getting memory errors. The query returns about 200 rows and fails after a few dozen, sometimes in exchange.php and sometimes in an actual exchange (bittrex.php, binance.php, etc), and always at a different line. I put verbose mode on and saw that it seems to be querying the exchange every single time.

Is there a cleaner solution?

`public function updateCoinPrice($where="") {
		$where = ($where == '' ? '':" AND $where");

		$myCoins = $this->conn->prepare("SELECT CoinID, CMCID, Coin, Currency, lcase(Exchange) AS Exchange FROM (trade_detail INNER JOIN list_coin ON trade_detail.CoinID = list_coin.ID) INNER JOIN list_exchange ON trade_detail.ExchangeID = list_exchange.ID INNER JOIN list_currency ON trade_detail.CurrencyID=list_currency.ID WHERE CurrencyID IN (1,2) $where GROUP BY list_coin.ID, CMCID, Coin, Currency, Exchange ORDER BY RAND()");
		$myCoins->execute();
		
		while ($row = $myCoins->fetch(PDO::FETCH_ASSOC)) {
			//Only update the coin once
			if (!$coins[$row['Coin'].$row['Currency']]) {
				
				writeLog($row['Exchange'] . ": " . $row['Coin']. "/" .$row['Currency']);
					
				try {				
					switch($row['Exchange']) {
						case "binance":
							$exchange = new \ccxt\binance (array ('verbose' => true)); break;
						case "bittrex":
							$exchange = new \ccxt\bittrex (array ('verbose' => true)); break;
						case "poloniex":
							$exchange = new \ccxt\poloniex (array ('verbose' => true)); break;
						case "yobit":
							$exchange = new \ccxt\yobit (array ('verbose' => true));  break;
						case "cryptopia":
							$exchange = new \ccxt\cryptopia (array ('verbose' => true)); break;
						case "kucoin":
							$exchange = new \ccxt\kucoin (array ('verbose' => true)); break;
						case "hitbtc":
							$exchange = new \ccxt\hitbtc (array ('verbose' => true)); break;
						case "liqui":
							$exchange = new \ccxt\liqui (array ('verbose' => true)); break;
						case "nova":
							$exchange = new \ccxt\nova (array ('verbose' => true)); break;
						case "livecoin":
							$exchange = new \ccxt\livecoin (array ('verbose' => true)); break;
						case "kucoin":
							$exchange = new \ccxt\kucoin (array ('verbose' => true)); break;
						default:
							sendTelegram("EXCHANGE NOT LISTED: " . $row['Exchange']);
							writeLog("EXCHANGE NOT LISTED: " . $row['Exchange']);
							break;							
					}				
		
					$data= $exchange->fetch_ticker($row['Coin'] . "/" . $row['Currency']);					
					$last = $data["bid"];				
					
					switch ($row['Currency']) {
					case "BTC":
						$USD = $last * $this->BTC_USD;
						break;
					case "ETH":
						$USD = $last * $this->ETH_USD;
						break;
					}					
						
					if ($last > 0) {
						$sql = "UPDATE list_coin SET DateUpdated=NOW(), " . $row['Currency'] . "=$last, USD=$USD WHERE ID=" . $row['CoinID'];
						writeLog("UPDATECOINPRICE: $sql");
							
						$stmt = $this->conn->prepare($sql);
						$stmt->execute();					
						
						$coins[$row['Coin'].$row['Currency']] = true;														
					}					
				} catch (Exception $e) {
					writeLog("[Caught Exception] " . $e->getMessage());
				}
				
				unset($exchange);
				unset($stmt);
			}
		}
	}`

**Comment by kroitor** [comment]
There's no wonder why you're getting memory errors. You're instantiating one exchange instance per loop iteration. This is not how you should do things. Instead of selecting all coins and then creating an exchange for each coin (you're basically duplicating tons of exchanges for coins from same exchange ids), you should:

1. select all exchange ids
2. pre-create exchange instances by instantiating just one exchange instance per exchange id
3. don't forget to enable rate limiting, as [described in the Manual](https://github.com/ccxt/ccxt/wiki/Manual#rate-limit)
4. load markets for each exchange, as [described in the Manual](https://github.com/ccxt/ccxt/wiki/Manual#loading-markets)
5. save exchanges instances to an assoc array that would be accessible later in your loop
6. select all coins
7. traverse each coin, using the pre-created exchange instance (don't recreate it again)

Basically, you show lack of understanding of basic OOP principles like *object lifetime* and *references to objects vs copies of objects*. I'd suggest doing more tutorials on object-oriented programming in PHP:

- https://www.youtube.com/results?search_query=php+oop
- https://www.google.ru/search?q=php+oop
- https://code.tutsplus.com/tutorials/object-oriented-php-for-beginners--net-12762
- http://php.net/manual/en/language.oop5.php

Let us know if does not answer your question.

**Comment by presence35** [comment]
Thank you. I'll look into that material.

**Comment by kroitor** [comment]
@presence35 yep, it's in your best interest. On the other hand, if you don't care about proper design principles, then this may work for you (not tested):

```PHP
public function updateCoinPrice($where="") {
		$where = ($where == '' ? '':" AND $where");

		$myCoins = $this->conn->prepare("SELECT CoinID, CMCID, Coin, Currency, lcase(Exchange) AS Exchange FROM (trade_detail INNER JOIN list_coin ON trade_detail.CoinID = list_coin.ID) INNER JOIN list_exchange ON trade_detail.ExchangeID = list_exchange.ID INNER JOIN list_currency ON trade_detail.CurrencyID=list_currency.ID WHERE CurrencyID IN (1,2) $where GROUP BY list_coin.ID, CMCID, Coin, Currency, Exchange ORDER BY RAND()");
		$myCoins->execute();
		
		while ($row = $myCoins->fetch(PDO::FETCH_ASSOC)) {
			//Only update the coin once
			if (!$coins[$row['Coin'].$row['Currency']]) {
				
				writeLog($row['Exchange'] . ": " . $row['Coin']. "/" .$row['Currency']);
					
				try {				
					switch($row['Exchange']) {
						case "binance":
							$exchange = new \ccxt\binance (array ('verbose' => true)); break;
						case "bittrex":
							$exchange = new \ccxt\bittrex (array ('verbose' => true)); break;
						case "poloniex":
							$exchange = new \ccxt\poloniex (array ('verbose' => true)); break;
						case "yobit":
							$exchange = new \ccxt\yobit (array ('verbose' => true));  break;
						case "cryptopia":
							$exchange = new \ccxt\cryptopia (array ('verbose' => true)); break;
						case "kucoin":
							$exchange = new \ccxt\kucoin (array ('verbose' => true)); break;
						case "hitbtc":
							$exchange = new \ccxt\hitbtc (array ('verbose' => true)); break;
						case "liqui":
							$exchange = new \ccxt\liqui (array ('verbose' => true)); break;
						case "nova":
							$exchange = new \ccxt\nova (array ('verbose' => true)); break;
						case "livecoin":
							$exchange = new \ccxt\livecoin (array ('verbose' => true)); break;
						case "kucoin":
							$exchange = new \ccxt\kucoin (array ('verbose' => true)); break;
						default:
							sendTelegram("EXCHANGE NOT LISTED: " . $row['Exchange']);
							writeLog("EXCHANGE NOT LISTED: " . $row['Exchange']);
							break;							
					}				
		
					$data= $exchange->fetch_ticker($row['Coin'] . "/" . $row['Currency']);					
					$last = $data["bid"];				
					
					switch ($row['Currency']) {
					case "BTC":
						$USD = $last * $this->BTC_USD;
						break;
					case "ETH":
						$USD = $last * $this->ETH_USD;
						break;
					}					
						
					if ($last > 0) {
						$sql = "UPDATE list_coin SET DateUpdated=NOW(), " . $row['Currency'] . "=$last, USD=$USD WHERE ID=" . $row['CoinID'];
						writeLog("UPDATECOINPRICE: $sql");
							
						$stmt = $this->conn->prepare($sql);
						$stmt->execute();					
						
						$coins[$row['Coin'].$row['Currency']] = true;														
					}	

//-----------------------------------------------------------------------

             				unset($exchange);  // ←----- moving these two lines up here might help
		        		unset($stmt);  // ←----- moving these two lines up here might help

//-----------------------------------------------------------------------
				
				} catch (Exception $e) {
					writeLog("[Caught Exception] " . $e->getMessage());
				}
			}
		}
	}
```

**Comment by presence35** [comment]
That didn't work.

Anyway, I rewrote it to load the exchanges only once and it's fine now.  Thanks for your help.

One other non-related question. I can't seem to load the BAY/BTC price for Bittrex. I tried using BitBay/BTC and that doesn't work either. The symbol value when I dump fetch_markets is BAY, so I'm confused why it's not working.
